python/ray/data/tests/test_dataset_pipeline.py

```diff
@@ -263,45 +263,30 @@ def test_schema(ray_start_regular_shared):
     assert pipe.schema() == int
 
 
-def test_schema_read_is_not_disruptive(ray_start_regular_shared):
-    # Multiple datasets: access schema and then iter_datasets()
+def test_schema_peak(ray_start_regular_shared):
+    # Multiple datasets
     pipe = ray.data.range(6).window(blocks_per_window=2)
     assert pipe.schema() == int
     dss = list(pipe.iter_datasets())
     assert len(dss) == 3, dss
-    assert pipe.schema() == int
-
-    # Multiple datasets: iter_datasets() and then access schema
-    pipe = ray.data.range(6).window(blocks_per_window=2)
-    dss = list(pipe.iter_datasets())
-    assert len(dss) == 3, dss
-    assert pipe.schema() == int
+    with pytest.raises(RuntimeError):
+        pipe.schema()
 
-    # Only 1 dataset: access schema and then iter_datasets()
+    # Only 1 dataset
     pipe = ray.data.range(1).window(blocks_per_window=2)
     assert pipe.schema() == int
     dss = list(pipe.iter_datasets())
     assert len(dss) == 1, dss
-    assert pipe.schema() == int
-
-    # Only 1 dataset: iter_datasets() and then access schema
-    pipe = ray.data.range(1).window(blocks_per_window=2)
-    dss = list(pipe.iter_datasets())
-    assert len(dss) == 1, dss
-    assert pipe.schema() == int
+    with pytest.raises(RuntimeError):
+        pipe.schema()
 
-    # Empty datasets: access schema and then iter_datasets()
+    # Empty datasets
     pipe = ray.data.range(6).filter(lambda x: x < 0).window(blocks_per_window=2)
     assert pipe.schema() is None
     dss = list(pipe.iter_datasets())
     assert len(dss) == 3, dss
-    assert pipe.schema() is None
-
-    # Empty datasets: iter_datasets() and then access schema
-    pipe = ray.data.range(6).filter(lambda x: x < 0).window(blocks_per_window=2)
-    dss = list(pipe.iter_datasets())
-    assert len(dss) == 3, dss
-    assert pipe.schema() is None
+    with pytest.raises(RuntimeError):
+        pipe.schema()
```

Comment:
Can we also test the first dataset reference is cleared (so we aren't leaking it in the object store)?

Reply:
Yep, done.